File: gunhoflash/implement-trie-prefix-tree.py
class Trie:
  def __init__(self):
    self.hasValue = False
    self.subTrie = {}

  def insert(self, word: str) -> None:
    if len(word):
      firstLetter = word[0]
      subTrie = self.subTrie.setdefault(firstLetter, Trie())
      subTrie.insert(word[1:])
    else:
      self.hasValue = True

  def search(self, word: str) -> bool:
    if len(word):
      firstLetter = word[0]
      subTrie = self.subTrie.get(firstLetter)
      return subTrie is not None and subTrie.search(word[1:])
    else:
      return self.hasValue

  def startsWith(self, prefix: str) -> bool:
    if len(prefix):
      firstLetter = prefix[0]
      subTrie = self.subTrie.get(firstLetter)
      return subTrie is not None and subTrie.startsWith(prefix[1:])
    else:
      return self.hasValue or len(self.subTrie) > 0

File: gunhoflash/implement-trie-prefix-tree.ts
class Trie {
  hasValue: boolean;
  subTrie: Map<string, Trie>;

  constructor() {
    this.hasValue = false;
    this.subTrie = new Map<string, Trie>();
  }

  insert(word: string): void {
    if (word.length) {
      const firstLetter = word[0];
      const subTrie = this.subTrie.get(firstLetter) || new Trie();
      subTrie.insert(word.slice(1));
      this.subTrie.set(firstLetter, subTrie);
    } else {
      this.hasValue = true;
    }
  }

  search(word: string): boolean {
    if (word.length) {
      const firstLetter = word[0];
      const subTrie = this.subTrie.get(firstLetter);
      return !!subTrie && subTrie.search(word.slice(1));
    } else {
      return this.hasValue;
    }
  }

  startsWith(prefix: string): boolean {
    if (prefix.length) {
      const firstLetter = prefix[0];
      const subTrie = this.subTrie.get(firstLetter);
      return !!subTrie && subTrie.startsWith(prefix.slice(1));
    } else {
      return this.hasValue || this.subTrie.size > 0;
    }
  }
}
